### DISCUSSION TITLE
onOpenChange callback only fires on close in Radix dialog

### DISCUSSION THREAD

**Discussion by ehsaneona** [radix-ui/primitives#3300] [Help]
Hi, I'm facing an issue with the onOpenChange callback in the Dialog component from Radix UI. I’m managing the open state manually (without using the Trigger component), and the onOpenChange callback only fires when the dialog closes (false). It doesn't fire when the dialog opens (true).

Here's the code I’m working with:

`<Dialog.Root open={isOpen} onOpenChange={handleOpenChange}>
      <Dialog.Portal>
        <Dialog.Overlay />
        <Dialog.Content>
          <Dialog.Title>My Dialog</Dialog.Title>
          <Dialog.Close>Close</Dialog.Close>
        </Dialog.Content>
      </Dialog.Portal>
      <button onClick={() => setIsOpen(true)}>Open Dialog</button>
    </Dialog.Root>`

**Comment by MaximKondratev** [comment]
Same problem here. `onOpenChange` seems to be fired only if internal state is changed but never if controlled state is. 

I made [this sandbox](https://codesandbox.io/p/devbox/radix-no-on-open-change-3663g5) to check it and dialog calls `onOpenChange` only if I interact with Dialog.Close or with overlay. Opening dialog with "edit profile" and closing it with "save changes" does not fire `onOpenChange` (because it is not a Dialog.Trigger and not a Dialog.Close respectively, but just buttons with onClick)

I can see why this is done this way, but it creates problems too. 

Just for context: I came here because this behavior breaks [Vaul's nested drawers](https://vaul.emilkowal.ski/default#nested-drawers) animation for controlled drawers where it depends on onOpenChange being called. Here is my [issue](https://github.com/emilkowalski/vaul/issues/540) about it.

**Comment by Oskii2311** [comment]
I have the same problem using version `1.1.6`. Have you found any workaround?

**Reply by ehsaneona** [reply]
I read in the radix ui documentation that it seems to be intentional and is because we have access to state outside the component.

**Reply by FranCarstens** [reply]
@Oskii2311 , if you have a way to track open state you can implement a "temporary" read, `TODO: fix one day`, useEffect that fires on the change in that open state.
